src/anemoi/datasets/create/input/action.py

```diff
@@ -7,251 +7,234 @@
-    from .pipe import PipeAction
-    from .repeated_dates import RepeatedDatesAction
-
-    # from .data_sources import DataSourcesAction
-
-    assert isinstance(context, Context), (type, context)
-    if not isinstance(config, dict):
-        raise ValueError(f"Invalid input config {config}")
-    if len(config) != 1:
-        print(json.dumps(config, indent=2, default=str))
-        raise ValueError(f"Invalid input config. Expecting dict with only one key, got {list(config.keys())}")
-
-    config = deepcopy(config)
-    key = list(config.keys())[0]
-
-    if isinstance(config[key], list):
-        args, kwargs = config[key], {}
-    elif isinstance(config[key], dict):
-        args, kwargs = [], config[key]
-    else:
-        raise ValueError(f"Invalid input config {config[key]} ({type(config[key])}")
-
-    cls = {
-        "data_sources": DataSourcesAction,
-        "data-sources": DataSourcesAction,
-        "concat": ConcatAction,
-        "join": JoinAction,
-        "pipe": PipeAction,
-        "function": FunctionAction,
-        "repeated_dates": RepeatedDatesAction,
-        "repeated-dates": RepeatedDatesAction,
-    }.get(key)
-
-    if cls is None:
-        from ..sources import create_source
-
-        source = create_source(None, substitute(context, config))
-        return FunctionAction(context, action_path + [key], key, source)
-
-    return cls(context, action_path + [key], *args, **kwargs)
+    def __init__(self, config, *path):
+        self.config = config
+        self.path = path
+        assert path[0] in (
+            "input",
+            "data_sources",
+        ), f"{self.__class__.__name__}: path must start with 'input' or 'data_sources': {path}"
+
+
+class Concat(Action):
+    def __init__(self, config, *path):
+        super().__init__(config, *path, "concat")
+
+        assert isinstance(config, list), f"Value must be a dict {list}"
+
+        self.choices = []
+
+        for i, item in enumerate(config):
+
+            assert "dates" in item, f"Value must contain the key 'dates' {item}"
+            dates = item["dates"]
+            filtering_dates = DatesProvider.from_config(**dates)
+            action = action_factory({k: v for k, v in item.items() if k != "dates"}, *self.path, str(i))
+            self.choices.append((filtering_dates, action))
+
+    def __repr__(self):
+        return f"Concat({self.choices})"
+
+    def __call__(self, context, argument):
+
+        results = context.empty_result()
+
+        for filtering_dates, action in self.choices:
+            dates = context.matching_dates(filtering_dates, argument)
+            if len(dates) == 0:
+                continue
+            results += action(context, dates)
+
+        return context.register(results, self.path)
+
+
+class Join(Action):
+    def __init__(self, config, *path):
+        super().__init__(config, *path, "join")
+
+        assert isinstance(config, list), f"Value must be a list {config}"
```

Comment:
```suggestion
        assert isinstance(config, list), f"Value of Join Action must be a list, got: {config}"
```
I've been running into these kind of errors. I believe being more verbose would help future-me.